### ISSUE TITLE
tensorflow.keras printing control characters in progress bars

### ISSUE THREAD

**Issue by austinmw** [tensorflow/tensorflow#24328]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 16.04, docker hub tag: 1.12.0-gpu-py3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): docker pull
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5
- Bazel version (if compiling from source): not source
- GCC/Compiler version (if compiling from source): not source
- CUDA/cuDNN version: 7.1.4
- GPU model and memory: P80 12GB

**Describe the current behavior**
I'm running tensorflow.keras through a docker container and displaying the output in a Jupyter notebook that is outside of the container (using Amazon SageMaker notebooks). Regular Keras printed progress bars correctly, but switching to tensorflow.Keras now spams my screen with control characters when trying to print progress bars. For example when Keras is downloading ImageNet weights for a model it outputs this:

```
Downloading data from https://github.com/fchollet/deep-learning-models/releases/download/v0.4/xception_weights_tf_dim_ordering_tf_kernels_notop.h5
#015    8192/83683744 [..............................] - ETA: 12:27#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015   40960/83683744 [..............................] - ETA: 5:00 #010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015   90112/83683744 [..............................] - ETA: 3:24#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015  212992/83683744 [..............................] - ETA: 1:55#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015  442368/83683744 [..............................] - ETA: 1:09#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015  892928/83683744 [..............................] - ETA: 40s #010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010#015 1810432/83683744 [..............................] - ETA: 23s#010#010#010#010#010#010#010#010#010#010#010#010#010#010#010
```

**Describe the expected behavior**
It shouldn't print control characters

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Reproducing this exactly may require running through SageMaker. The example I modified is this: https://github.com/awslabs/amazon-sagemaker-examples/tree/master/hyperparameter_tuning/keras_bring_your_own 
Specifically, I changed all of the Keras imports in this file: https://github.com/awslabs/amazon-sagemaker-examples/blob/master/hyperparameter_tuning/keras_bring_your_own/trainer/start.py to be tensorflow.keras imports. 

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by austinmw** [comment]
Hi, thanks for your response. I'm running Python 3.5.x with both. I've put the model code in a tensorflow docker image and uploaded it to the AWS ECR. Then when I run the SageMaker.estimator.Estimator.fit command while inside a SageMaker notebook instance it spins up an other EC2 instance, pulls the ECR image I uploaded, runs the model on there, then displays the output in the SageMaker notebook instance I called the fit command from.

**Comment by AnakTeka** [comment]
> Hi, thanks for your response. I'm running Python 3.5.x with both. I've put the model code in a tensorflow docker image and uploaded it to the AWS ECR. Then when I run the SageMaker.estimator.Estimator.fit command while inside a SageMaker notebook instance (passing it the ECR path, an instance type, and an S3 train/valid data path) it spins up a separate EC2 instance, pulls the ECR image to it, pulls the data from S3, trains the model on there (and saves model artifacts to S3), while displaying the training print outputs in the SageMaker notebook instance cell from which I originally called the fit command.
> 
> All-in-all a difficult process for me to track, but what I do know is that this issue began when I switched from keras to tensorflow.keras.

I'm having a same problem, did you find the fix for this?

**Comment by austinmw** [comment]
> > Hi, thanks for your response. I'm running Python 3.5.x with both. I've put the model code in a tensorflow docker image and uploaded it to the AWS ECR. Then when I run the SageMaker.estimator.Estimator.fit command while inside a SageMaker notebook instance (passing it the ECR path, an instance type, and an S3 train/valid data path) it spins up a separate EC2 instance, pulls the ECR image to it, pulls the data from S3, trains the model on there (and saves model artifacts to S3), while displaying the training print outputs in the SageMaker notebook instance cell from which I originally called the fit command.
> > All-in-all a difficult process for me to track, but what I do know is that this issue began when I switched from keras to tensorflow.keras.
> 
> I'm having a same problem, did you find the fix for this?

Nope.

**Comment by uchua** [comment]
Just started encountering this issue when using TensorFlow 2.0.0 and 2.1.0 in SageMaker. Is there still no fix for this?

**Comment by aninoy** [comment]
I'm also facing this issue. Has anyone figured out a solution yet?

**Comment by ishaaq** [comment]
I think the issue is here:
https://github.com/tensorflow/tensorflow/blob/bbb81ea42831b244289a059de9e4be0203dc4c35/tensorflow/python/keras/utils/generic_utils.py#L337-L340

The code running inside SageMaker neither has a tty nor is running inside jupyter, yet the check above enables `_dynamic_display` because of the `'posix' in sys.modules` check. I don't understand why that check exists.

Enabling the flag causes the backspace chars to be emitted in a platform that doesn't actually have an interactive console that supports it:
https://github.com/tensorflow/tensorflow/blob/bbb81ea42831b244289a059de9e4be0203dc4c35/tensorflow/python/keras/utils/generic_utils.py#L387-L389

**Comment by sonebu** [comment]
Not really a solution but a compromise: if you don't have to see the progress within each epoch you can use `model.fit(..., verbose=2)` . No annoying control characters this way so it's easier to navigate the output. From the documentation (11.05.2020):

> verbose: 0, 1, or 2. Verbosity mode. 0 = silent, 1 = progress bar, 2 = one line per epoch. Note that the progress bar is not particularly useful when logged to a file, so verbose=2 is recommended when not running interactively (eg, in a production environment).

That "not running interactively" part seems to be the key here

**Comment by blink1073** [comment]
There is a relevant discussion about the handling of backspace characters in Jupyter Notebook and JupyterLab [here](https://github.com/jupyter/notebook/issues/5381).

**Comment by roccopietrini** [comment]
same issue in AWS Sagemaker with tensorflow version 2.3.0.. any news?

**Comment by pritesh-aidash** [comment]
I'm facing the same issue. Can someone please shed some light here.

**Comment by aninoy** [comment]
> I'm facing the same issue. Can someone please shed some light here.

I was able to resolve this by using `verbose=2` in the fit method. https://keras.io/api/models/model_training_apis/

**Comment by rmothukuru** [comment]
@austinmw,
Can you please confirm if above comment helped you to resolve your issue? Thanks!

**Comment by ishaaq** [comment]
Can't speak for @austinmw but I can say that, while setting `verbose=2` certainly solves the problem in SageMaker specifically but also jupyter consoles in general, as @sonebu mentioned above, it's only a compromise as it solves by simply not emitting progress bars.

The real question here is why the code is treating a non-tty-capable output as a tty-capable output as mentioned in my comment above: https://github.com/tensorflow/tensorflow/issues/24328#issuecomment-616952828

**Comment by deeb02** [comment]
This behaves as intended. The default output is intended for typical terminals and notebook environments. For other types of environments and logging please use verbose=0 or verbose=2.
